payjoin/src/receive/mod.rs

```diff
@@ -325,25 +325,114 @@ impl OutputsUnknown {
             return Err(Error::BadRequest(InternalRequestError::MissingPayment.into()));
         }
 
-        Ok(ProvisionalProposal {
-            original_psbt: self.psbt.clone(),
-            payjoin_psbt: self.psbt,
-            params: self.params,
-            owned_vouts,
-        })
+        Ok(WantsOutputs { psbt: self.psbt, params: self.params, owned_vouts })
     }
 }
 
-/// A mutable checked proposal that the receiver may contribute inputs to to make a payjoin.
+/// A checked proposal that the receiver may substitute or add outputs to
 #[derive(Debug, Clone)]
-pub struct ProvisionalProposal {
+pub struct WantsOutputs {
+    psbt: Psbt,
+    params: Params,
+    owned_vouts: Vec<usize>,
+}
+
+impl WantsOutputs {
+    pub fn is_output_substitution_disabled(&self) -> bool {
+        self.params.disable_output_substitution
+    }
+
+    /// If output substitution is enabled, replace the receiver's output script with a new one.
+    pub fn try_substitute_receiver_output(
+        self,
+        generate_script: impl Fn() -> Result<bitcoin::ScriptBuf, Error>,
+    ) -> Result<WantsInputs, Error> {
+        let output_value = self.psbt.unsigned_tx.output[self.owned_vouts[0]].value;
+        let outputs =
+            vec![(TxOut { value: output_value, script_pubkey: generate_script()? }, true)];
+        self.try_substitute_receiver_outputs(Some(outputs))
+    }
+
+    pub fn try_substitute_receiver_outputs(
+        self,
+        outputs: Option<Vec<(TxOut, bool)>>, // (txout, is_owned)
```

Comment:
This tuple vs option construction confuses me a bit since I assumed there would only ever be one receiver change `is_owned` which fees were taken from. I suppose there could be multiple if one change were completely exhausted but that scenario seems overly complex vs guiding someone downstream to pick one (or no) change output.

I recognize you've put a ton of effort into this design and also want to document another line of thinking we started exploring in Nashville to come up with a design. It's a bit of a tangent but bear with me:

We had discussed a builder pattern and I wonder if that would give feedback to a branch-and-bound *output* selection algorithm to pick from e.g. an exchange's queue of outgoing transactions. Once `WantsOutputs` was created, which would remove the receiver outputs from the payjoin_psbt, Each call to `try_add_output(TxOut) -> Result<WantsInputs, Error>` would check the rules to see if a single output could be added and return a `` that an iterator could use to keep adding outputs until it added a change. `WantsOutputs::excess_input() -> Amount` could let a caller check how much they have to spend. Then, finally a `build(change_vout) -> WantsInputs` could do final checks / shuffling. build could take options to potentially sweep (no change), have multiple change, or parameterize a shuffling. Perhaps there were limitations in this thinking or we hadn't fully explored the idea then.

Reply:
Fair enough re: multiple change outputs being overly complex.

I'm not particularly attached to this current design and happy to explore the builder pattern approach. I vaguely recall it was brought up in Nashville but don't remember any specifics. I have questions on the approach you described above.

> Once `WantsOutputs` was created, which would remove the receiver outputs from the payjoin_psbt

What if the receiver doesn't want to substitute outputs and wants to keep the original?

> Each call to `try_add_output(TxOut) -> Result<WantsInputs, Error>` would check the rules to see if a single output could be added and return a `` that an iterator could use to keep adding outputs until it added a change

- If this is meant to be called multiple times following a builder pattern shouldn't it just mutate `WantsOutputs` instead of returning a `WantsInputs`?
- What rules are we checking here? As far as I understand there are no limitations to what outputs can be added (perhaps if `disable_output_substitution` is true, but that seems like a `build()` check).
- What return type did you mean here: "return a `` that an iterator could use" ?

> WantsOutputs::excess_input() -> Amount could let a caller check how much they have to spend

Based only on the sender input? IIUC this is all happening before the receiver contributes their input(s).

> `build(change_vout) -> WantsInputs`

I don't think we can count on the caller to pass `change_vout`, since this step also shuffles outputs ordering. It seems to me they would have to pass a `&TxOut` or script pubkey, that the `build()` method then uses to identify the correct `vout`.

Reply:
My comment was a bit reckless since I wanted to give you some feedback in haste, and I definitely made a few mistakes. Thank you for catching them.

> > Each call to `try_add_output(TxOut) -> Result<WantsInputs, Error>` would check the rules to see if a single output could be added and return a `` that an iterator could use to keep adding outputs until it added a change
> 
> - If this is meant to be called multiple times following a builder pattern shouldn't it just mutate `WantsOutputs` instead of returning a `WantsInputs`?

yes

> - What rules are we checking here? As far as I understand there are no limitations to what outputs can be added (perhaps if `disable_output_substitution` is true, but that seems like a `build()` check).

disable_output_substitution was what I was thinking since that's a separate code path from the point of view of coin selection code

> - What return type did you mean here: "return a `` that an iterator could use" ?

I think I edited my comment incompletely. I was starting to address the `excess_input()` function that would return the current state of sender input minus defined outputs so that downstream could loop on the result of that function returning the sate of `WantsOutput` until output was satisfied based just on the sender input to do as pure a cut-through as possible without contributing receiver input. This may not actually be a worthwhile api complication if we want to discourage 0-receiver-input cut-through.

> I don't think we can count on the caller to pass change_vout, since this step also shuffles outputs ordering. It seems to me they would have to pass a &TxOut or script pubkey, that the build() method then uses to identify the correct vout.

You're definitely correct.